manga109api/manga109api.py

```diff
@@ -17,6 +19,66 @@ def __init__(self, root_dir):
         with (self.root_dir / "books.txt").open("rt", encoding='utf-8') as f:
             self.books = [line.rstrip() for line in f]
 
+
+    def get_annotation_ordered(self, book, annotation_type="annotations"):
+        """
+        Given a book title, return the annotation in the form of a dict.
+
+        In the raw XML in the Manga109 dataset, the bounding box data in the
+        `page` tag is not sorted by its annotation type, and each bounding
+        box type appears in an arbitrary order. The output dictionary of this
+        function preserves the ordering of each bounding box tag in the raw
+        XML data, mainly for data editing purposes. Note that the ordering of
+        the bounding box tags does not carry any useful information about the
+        contents of the data.
+
+        Caution: Due to the aforementioned feature, the format of the output
+        dictionary differs from the `get_annotation` function.
+
+        Args:
+            book (str): A title of a book. Should be in self.books.
+
+        Returns:
+            annotation (dict): Annotation data consists of dict.
+        """
+        assert book in self.books
+
+        def keys_renamed(d):
+            for k in list(d.keys()):
+                try:
+                    if str(d[k]) == str(int(d[k])):
+                        d["@"+k] = int(d.pop(k))
+                    else:
+                        d["@"+k] = d.pop(k)
+                except ValueError as e:
+                    d["@"+k] = d.pop(k)
```

Comment:
'else' case and 'except' case do same thing `d["@"+k] = d.pop(k)`
thus, we should remove one of them? (e.g. remove 'if' & 'else' and only use try/except)

Reply:
I thought for a moment too for simplifying this part too. The complicated part is that this function here is actually trying to capture three patterns:

- (1): The `str(int(x))` throws an exception and fails, since `x` cannot be interpreted as an integer literal, such as when `x == "some text"` . In this case, return `x`.
- (2): The `str(int(x))` succeeds, but `if str(x) == str(int(x))` is False, such as when `x == "00001234"`. In this case, return `x`.
- (3): The `str(int(x))` succeeds, and `if str(x) == str(int(x))` is True. In this case, return `int(x)`.

In the new code here https://github.com/manga109/manga109api/blob/add-ordered-output-func/manga109api/manga109api.py#L51 , I chose to throw a `ValueError` in the exception in the (2) case. Since (1) also throws a ValueError (when `str(int(x))`, or more precisely, when `int(x)` fails), this can capture both cases. Case (3) is handled inside the exception.

One concern was whether if the `ValueError` exception is a natural choice to `throw` in the `except` block. What are your thoughts about this and the overall code?

Reply:
I see. I understand your purpose now :)

> One concern was whether if the ValueError exception is a natural choice to throw in the except block.

I feel slightly unnatural because (`if str(x) == str(int(x))` is False) is not ValueError?

In this case, in my opinion, the previous code was easy to understand. (sorry for bothering you)
I needed simple comments/examples to understand the purpose.
```
            try:
                if str(t) == str(int(t)):
                    return int(t)
                else:
                    return t    # for the case, such as t = "00001234"
            except ValueError as e:
                return t     # for the case, such as t = "some text" or t = "000012ab"
```

Reply:
Thanks, yes, I was feeling that raising a ValueError was a little unnatural too. I'm glad the previous code was easy to understand, no problem at all, I'll put it back and also put some comments as in your suggestion.